### PR TITLE
[2/8] Big merge - core

### PR DESCRIPTION
**Pull request by tscmoo**
A lot of code has moved to core from being poorly located in mcts and tube.

core::Actor is now a base class separated from MCTS.
ActorPlayer is a player backed by Actor, mostly used as a base for MctsPlayer and ForwardPlayer.
ForwardPlayer is a new player used for non-mcts things which just passes inputs/outputs directly to/from the actor (and thus the model).
The base Player class moved here from MCTS.

ModelManager is a refactor of ChannelAssembler from tube, as ChannelAssembler had grown out of scope of its original design. ModelManagers main purpose is to hold and run the model (for training). It also has some new code for finding the optimal batch size as used by workers.

The replay buffer moved here from MCTS and had a complete rewrite to improve some performance issues. Behavior should be the same as before.

State moved into the core namespace and had some (performance) improvements.
  - _Action is now not meant to be derived from. It thus no longer has virtual functions, and it is stored directly instead of through shared_ptr.
  There are two new functions clearActions() and addAction(x, y, z) which should be used to fill in the actions. If a game needs to store more information related to an action, they must do so internally in their State implementation, for instance in an array/vector which can be looked up by the index of the action (which will always be set in the order actions were added).
  - State implementations must now be copy-assignable (`State& operator=(State&)`). The default assignment operator works for most games, as long as they do not have const members.
  - FeatureOptions now stores the feature options, instead of them being static members. setFeatures must be called with a pointer to FeatureOptions. This object must outlive State itself. Game takes care of this normally.
  

Game has some new features and refactoring, but overall it grows larger, more complex and out of hand. I intent to refactor much of the logic and extract features out into separate classes which will interact with games through a common interface - but for now it is as it is.

- In some proportion of games, it will sometimes(/rarely) make a completely random move.
- In addition, on occasion, games will start with a series of random moves.
- There is support for logit value outputs from the model (ie a softmax output of win/loss/draw), though no models currently implement this. The value backpropagated to the mcts will be the probabilities of `win - loss` resulting in a range of [-1, 1] with 0 being draw. The MCTS, however, has no support yet for backpropagating the draw probability by itself. This probability could be used to implement resigning & automatic draw.
- There is experimental support for RNN models. This implementation unfortunately still has some issues when used with MCTS (but will not be too hard to fix in the future). There are, however, no RNN models yet.
- There is very experimental support for non-MCTS RL algorithms, though no suitable loss functions 
are implemented yet.
- There is also support for training against another model. This model can have a different architecture and featurization, and can even run a different game implementation, as long as the action space is shared. I'm not entirely sure about the usefulness of this right now, beyond speeding up training by training against good models from other runs.
- There is a new featurization turnFeaturesMultiChannel which does the same as the old turnFeatures (now turnFeaturesSingleChannel) - this has one channel per player, set to 1 if it is that players turn, and 0 otherwise. The reason is just that with only a single channel, there is a small bias near the edge of the board caused by conv padding.
- There is support for side learning through predicting the next N game states. It's also possible to target the last game state (predicting the end)
- There is a new "rewind" feature, which will, on game end, rewind the game to the last point where the model evaluated the game state as the opposite of the final outcome (effectively starting a new game from that game state). It will do this a specified number of times before starting a new game like normal.

There is also some new multithreading, with the idea being that any CPU processing should happen in small jobs which can be ran in parallel, and there is a fixed pool of threads that do this processing. game.cc has been partially updated to use this for stepping through games, but it still starts and ends games sequentially (this will be updated to be parallel later). MCTS uses this multithreading quite well.

### CODE REVIEW

**Comment by tscmoo** [comment]
Merged in https://github.com/facebookincubator/Polygames/commit/b23f1c83b1741bedff99ef88e4b12240abd87f03
